src/optimize/bundles_route/dynamic_asset_response.ts

```diff
@@ -51,48 +64,68 @@ import { replacePlaceholder } from '../public_path_placeholder';
  *  @property {string} options.publicPath
  *  @property {LruCache} options.fileHashCache
  */
-export async function createDynamicAssetResponse(options) {
-  const { request, h, bundlesPath, publicPath, fileHashCache, replacePublicPath } = options;
-
-  let fd;
+export async function createDynamicAssetResponse({
+  request,
+  h,
+  bundlesPath,
+  publicPath,
+  fileHashCache,
+  replacePublicPath,
+  isDist,
+}: {
+  request: Hapi.Request;
+  h: Hapi.ResponseToolkit;
+  bundlesPath: string;
+  publicPath: string;
+  fileHashCache: LruCache<unknown, unknown>;
+  replacePublicPath: boolean;
+  isDist: boolean;
+}) {
+  let fd: number | undefined;
   try {
     const path = resolve(bundlesPath, request.params.path);
 
     // prevent path traversal, only process paths that resolve within bundlesPath
     if (!path.startsWith(bundlesPath)) {
-      throw Boom.forbidden(null, 'EACCES');
+      throw Boom.forbidden(undefined, 'EACCES');
     }
 
     // we use and manage a file descriptor mostly because
     // that's what Inert does, and since we are accessing
     // the file 2 or 3 times per request it seems logical
-    fd = await fcb(cb => open(path, 'r', cb));
+    fd = await asyncOpen(path, 'r');
 
-    const stat = await fcb(cb => fstat(fd, cb));
-    const hash = await getFileHash(fileHashCache, path, stat, fd);
+    const stat = await asyncFstat(fd);
+    const hash = isDist ? undefined : await getFileHash(fileHashCache, path, stat, fd);
 
-    const read = createReadStream(null, {
+    const read = Fs.createReadStream(null as any, {
       fd,
       start: 0,
       autoClose: true,
     });
-    fd = null; // read stream is now responsible for fd
+    fd = undefined; // read stream is now responsible for fd
 
     const content = replacePublicPath ? replacePlaceholder(read, publicPath) : read;
-    const etag = replacePublicPath ? `${hash}-${publicPath}` : hash;
 
-    return h
+    const response = h
       .response(content)
       .takeover()
       .code(200)
-      .etag(etag)
-      .header('cache-control', 'must-revalidate')
       .type(request.server.mime.path(path).type);
+
+    if (isDist) {
+      response.header('cache-control', `max-age=${365 * DAY}`);
```

Comment:
Have we considered using `stale-while-revalidate` on top of `max-age`? 
It is _only_ supported by modern browsers, but I think it will benefit the users that are actually running them (without affecting others not using them). 
The idea behind it is: even when the cached resource is stale, the browser is allowed to use it but it will trigger a background network revalidation.

Reply:
I'm not totally opposed, but sounds unnecessary and like it might give us a false sense of security. The assets that we serve from our unique urls should never change so there shouldn't be any need to revalidate.

Reply:
I must say I'm also concerned about `max-age` being equal to 1 year. Wouldn't that cause issues for upgraded deployments?
Users will be forced to _hard-refresh_ the cache. And it's not as simple as Shift+F5 or Ctrl/Cmd+Shift+R since it enforces the first assets to be refreshed but not the ones down the tree (just tested it locally):
![image](https://user-images.githubusercontent.com/5469006/80477515-f685cf00-8943-11ea-9f90-edd52d69ff03.png)
_Mind the first-level assets are downloaded again (200), but the rest are receiving 304 (not forced refresh by the browser)_

Maybe our release cadence is a more suitable value? Should we take into consideration the patch-release cadence or the minor one? 🤔 

This is our version (minor) adoption, according to our telemetry:
![image](https://user-images.githubusercontent.com/5469006/80478572-9a23af00-8945-11ea-927c-c6e621177a25.png)

It looks like users running the latest version are more likely to upgrade to the next minor when it is released.

Reply:
Oh! I totally missed the `buildHash` is now part of the path in the requests! Dismiss my comment then :)

Reply:
> it might give us a false sense of security.

It's not about security. It's about speed. The browser is allowed to use the stale version of the cache during this render time, but it will revalidate the cache in the background.

i.e.: With `Cache-Control: stale-while-revalidate=30days; max-age=365days` the browser is allowed to use the cached asset during 365 days. One year later, it will set the asset as stale. If the asset is requested in the following 30days (say in day 367), the browser will still use the cached asset to render the site, but it will perform the network request to retrieve the asset in the background.

Reply:
Maybe a better alternative (since we have the `bundleHash` now in the path), is the property `inmutable`. Check **Caching static assets** in https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control

Reply:
> It's not about security. It's about speed.

I meant a false sense of security that the cache was being invalidated properly as in some cases it wouldn't be updated when we expect it to but a simple refresh would solve it, when in reality we want the right assets to be served the first time every time and for browsers to never get stale assets.